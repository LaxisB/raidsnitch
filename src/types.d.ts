type HandleTypes = FileSystemDirectoryHandle | FileSystemFileHandle;

interface FileSystemFileHandle extends FileSystemHandle {
    getFile(): Promise<File>;
    createSyncAccessHandle(): Promise<FileSystemSyncAccessHandle>;
}
interface FileSystemSyncAccessHandle {
    close(): Promise<undefined>;
    flush(): Promise<undefined>;
    getSize(): Promise<number>;
    read(buffer: Buffer, opts?: FileSystemReadWriteOptions): Promise<number>;
}
interface FileSystemDirectoryHandle extends FileSystemHandle {
    entries(): AsyncIterator<[string, HandleTypes]>;
    values(): AsyncIterator<HandleTypes>;
    keys(): AsyncIterator<string>;
}

interface FileSystemReadWriteOptions {
    /** offset to start reading from */
    at: number;
}

interface AsyncIterator<T, TReturn, TNext> {
    [Symbol.asyncIterator](): any;
}

interface FileSystemHandle {
    queryPermission(): Promise<'granted' | 'denied' | 'prompt'>;
    requestPermission(opts?: { mode: 'read' | 'readwrite' }): Promise<'granted' | 'denied'>;
}

interface Window {
    showDirectoryPicker: () => Promise<FileSystemDirectoryHandle>;
    showOpenFilePicker: (opts?: { multiple: boolean }) => Promise<FileSystemFileHandle[]>;
    showSaveFilePicker: (opts?: { suggestedName: string }) => Promise<FileSystemFileHandle>;
}
