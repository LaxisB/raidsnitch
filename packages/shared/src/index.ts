export * as log from './log';
export * as emitter from './emitter';
export * as rpc from './rpc';
export * as utils from './utils';
export * as format from './format';
