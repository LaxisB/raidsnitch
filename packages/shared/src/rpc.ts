import { wrapLog } from './log';

const log = wrapLog('rpc');
let requestCounter = 0;

function generateId() {
  return requestCounter++;
}

/**
 * create a handler function that dispatches incoming RPC requests to it's handlers
 * @param handlers Record of method names to register with their handler functions
 * @param middleware Array of middleware functions to run on each request
 * @param internalCb adapter for the transport layer (e.g. postMessage)
 */
export function setupRpcHandlers(handlers: Record<string, Function>, internalCb: any) {
  return function (request: RpcRequest<unknown>) {
    if (handlers[request.method]) {
      try {
        const res = handlers[request.method](...((request.params as any) ?? []));
        if (request.id == null) {
          // request was 'notification' style, so we don't need to respond
          return;
        }
        if ('then' in res) {
          return res
            .then((result: any) => {
              internalCb(request, { result });
            })
            .catch((error: any) => {
              internalCb(request, { error });
            });
        }
        internalCb(request, { result: res });
      } catch (e) {
        internalCb(request, { error: e });
      }
    } else {
      log.warn('unhandled request', request);
    }
  };
}

type InternalCb<T> = (req: RpcRequest<any>, transferArgs?: number[]) => Promise<T>;

/**
 * wraps a function to be callable via RPC
 */
export function asRpcCall<Fn extends (...args: any) => any>(name: string, internalCb: InternalCb<ReturnType<Fn>>) {
  return async function (...data: Parameters<Fn>): Promise<Awaited<ReturnType<Fn>>> {
    const request = {
      type: 'rpc' as 'rpc',
      method: name,
      params: data,
      id: generateId(),
    };

    return internalCb(request) as any;
  };
}

/**
 * create a RPC client that can be used to call methods on the server
 * it is a naive implementation that assumes that the server will respond to all requests
 */
export function createNaiveClient<T, B extends {}>(baseObject: B, worker: Worker) {
  const pendingRpcCalls = new Map<number, { resolve: (value: any) => void; reject: (reason?: any) => void }>();
  const post: any = (data: any) => {
    return new Promise((resolve, reject) => {
      log.debug('request', data.id, data.method, data.params);
      pendingRpcCalls.set(data.id, { resolve, reject });
      worker.postMessage(data);
    });
  };

  worker.addEventListener('message', (e) => {
    const res = e.data as RpcResponse<any, any>;
    if (res.type !== 'rpc') {
      return;
    }
    if (pendingRpcCalls.has(res.id)) {
      const { resolve, reject } = pendingRpcCalls.get(res.id)!;
      pendingRpcCalls.delete(res.id);
      if ('error' in res) {
        log.warn('error response', res);
        reject(res.error);
      } else {
        log.debug('response', res.id, res.result);
        resolve(res.result);
      }
    }
  });

  let fnCache: Record<string | symbol, any> = {};
  const client = new Proxy(baseObject, {
    get: function (target, prop, _receiver) {
      if (prop in target) {
        return (target as any)[prop];
      }
      if (fnCache[prop]) {
        return fnCache[prop];
      }
      fnCache[prop] = asRpcCall(prop as string, post);
      return fnCache[prop];
    },
  });
  return client as B & AsyncClient<T>;
}

export type AsyncClient<T> = {
  [K in keyof T]: T[K] extends (...args: infer A) => infer R ? (...args: A) => AsyncResponse<R> : never;
};

type AsyncResponse<T> = T extends Promise<unknown> ? T : Promise<T>;

export interface RpcRequest<T> {
  type: 'rpc';
  method: string;
  params: T;
  id: number;
}

interface RpcResponseBase {
  type: 'rpc';
  id: number;
}
interface RpcErrorResponse<E> extends RpcResponseBase {
  error: {
    message: string;
    data: E;
  };
}
interface RpcSuccessResponse<T> extends RpcResponseBase {
  result: T;
}
export type RpcResponse<T, E> = RpcErrorResponse<E> | RpcSuccessResponse<T>;

export interface RpcEvent {
  type: 'event';
  event: string;
  data: any;
}
