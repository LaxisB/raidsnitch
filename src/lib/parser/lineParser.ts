const LIST_START = ['[', '('];
const LIST_END = [']', ')'];
const SEPARATORS = [',', ...LIST_START, ...LIST_END];

type LogBottomValues = null | number | string;
type LogValues = LogBottomValues | LogValues[];

export interface DataLine {
    time: 0;
    offset: number;
    event: string;
    payload: LogValues[];
    line: string;
}

export interface ErrorLine {
    time: 0;
    offset: number;
    event: null;
    payload: null;
    line: string;
}

export type LogLine = DataLine | ErrorLine;
export interface Parser {
    parseLine(line: string): LogLine;
}

export function createLineParser(referenceTime: number) {
    let reftime = referenceTime;
    let offset = 0;

    return {
        parseLine(line: string): LogLine {
            const [timeblock, rest] = line.split('  ');
            if (!timeblock || !rest) {
                return {
                    time: 0,
                    offset: offset++,
                    event: null,
                    payload: null,
                    line,
                } as ErrorLine;
            }

            const firstComma = rest.indexOf(',');
            const event = rest.slice(0, firstComma);
            const body = rest.slice(firstComma + 1);

            const time = parseTimeblock(timeblock, reftime);
            const payload = parsePayload(body);

            return {
                time,
                offset: offset++,
                event,
                payload,
                line,
            } as LogLine;
        },
    };
}

const TIMESTAMP_MATCHER = new RegExp(/(\d{1,2})\/(\d{1,2}) (\d{1,2}):(\d{1,2}):(\d{1,2}).(\d{1,3})/);
function parseTimeblock(timeBlock: string, reftime: number): number {
    const matches = TIMESTAMP_MATCHER.exec(timeBlock);
    const d = new Date(reftime);
    d.setMonth(parseInt(matches![1]) - 1);
    d.setDate(parseInt(matches![2]));
    d.setHours(parseInt(matches![3]));
    d.setMinutes(parseInt(matches![4]));
    d.setSeconds(parseInt(matches![5]));
    d.setMilliseconds(parseInt(matches![6]));
    return d.getTime();
}

function parsePayload(body: string) {
    const stack = makeStack();
    const payload = _parsePayload(body, stack);
    return payload as LogValues[];
}

function _parsePayload(rest: string, stack: Stack): any[] {
    // EOL :::: exit branch
    // this is the only
    // we're checking for length to avoid matching long strings
    if (!rest || (rest.length < 2 && /^\s*$/.test(rest))) {
        return stack.value();
    }

    // nil val
    if (rest.startsWith('nil')) {
        stack.addVal(null);
        return _parsePayload(rest.slice(4), stack);
    }

    // flags
    if (rest.startsWith('0x')) {
        const [val, rest2] = splitAtSeparator(rest);
        const flags = parseFlags(val);
        stack.addVal(flags);
        return _parsePayload(rest2, stack);
    }

    // quoted string
    if (rest.startsWith('"')) {
        // assume that the char following the closing quote is a separator
        // there is no precedent to indicate that this is not always the case
        const [val, rest2] = splitAtQuote(rest.slice(1));
        stack.addVal(val.slice(0, -1).trim());
        return _parsePayload(rest2, stack);
    }

    // lists
    if (LIST_START.includes(rest[0])) {
        stack.addLayer();
        return _parsePayload(rest.slice(1), stack);
    }
    if (LIST_END.includes(rest[0])) {
        stack.collapseLayer();
        // remove the separator and trailing comma
        const rest2 = rest.slice(1).replace(/^,/, '');
        return _parsePayload(rest2, stack);
    }

    // special case: a null guid
    if (rest.startsWith('0000000000000000')) {
        const [val, rest2] = splitAtSeparator(rest);
        stack.addVal(val);
        return _parsePayload(rest2, stack);
    }

    // value (int/float/const)
    const [val, rest2] = splitAtSeparator(rest);
    stack.addVal(parsePrimitive(val));
    return _parsePayload(rest2, stack);
}

/**
 * Parses a hex-formatted string of into an integer
 */
function parseFlags(flags: string) {
    if (!flags.startsWith('0x')) {
        return 0;
    }
    return Number.parseInt(flags.slice(2), 16);
}

function parsePrimitive(val: string) {
    const isNum = /^-?\d+(\.\d+)?$/.test(val);
    return isNum ? Number.parseFloat(val) : val;
}

function splitAtQuote(string: string) {
    const offset = string.indexOf('"');
    return [string.slice(0, offset + 1), string.slice(offset + 2)];
}

function splitAtSeparator(string: string) {
    const offsets = SEPARATORS.map((sep) => string.indexOf(sep)).filter((offset) => offset > -1);
    const min = Math.min(string.length, ...offsets);
    const val = string.slice(0, min);
    // trim leading comma if present. we don't need it for parsing
    const rest = string[min] === ',' ? string.slice(min + 1) : string.slice(min);
    return [val, rest];
}

interface Stack {
    addVal(val: any): void;
    addLayer(): void;
    collapseLayer(): void;
    value(): any[];
}

function makeStack(): Stack {
    const values: any[] = [[]];

    const head = () => values[values.length - 1];

    return {
        addVal(val: any) {
            head().push(val);
        },
        addLayer() {
            values.push([]);
        },
        collapseLayer() {
            const head = values.pop();
            values[values.length - 1].push(head);
        },

        value() {
            return values[0];
        },
    };
}
