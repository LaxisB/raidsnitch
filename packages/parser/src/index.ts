import { parseAsEvent } from './events/parser';
import { createLineParser } from './lineParser';

export type * from './events';
export { casts } from './events';

export type Parser = ReturnType<typeof createParser>;

export function createParser(referenceTime: number) {
  const lineParser = createLineParser(referenceTime);

  function parseLine(line: string) {
    const logLine = lineParser.parseLine(line);
    if (logLine.event === null) {
      return null;
    }
    return parseAsEvent(logLine, referenceTime);
  }

  return {
    parseLine,
  };
}
