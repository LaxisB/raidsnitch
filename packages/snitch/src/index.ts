import { WowEvent } from '@raidsnitch/parser';
import { createDpsHandler } from './handlers/dps';
import { createEntitiesHandler } from './handlers/entities';
import { createHpsHandler } from './handlers/hps';
import { createSegmentHandler } from './handlers/segments';
import { createZoneHandler } from './handlers/zone';

export type State = ReturnType<typeof initialize>['state'];
export interface Snitch {
  state: State;
  handleEvents(events: WowEvent[]): void;
}

export function initialize(handleStats: (stats: State) => void) {
  let refTime = -1;
  const zone = createZoneHandler();
  const entities = createEntitiesHandler();
  const dps = createDpsHandler();
  const hps = createHpsHandler();
  const segments = createSegmentHandler();

  let state = {
    zone: zone.initialState,
    entities: entities.initialState,
    dps: dps.initialState,
    hps: hps.initialState,
    segments: segments.initialState,
  };

  const handlers = [segments.handleEvent, zone.handleEvent, entities.handleEvent, dps.handleEvent, hps.handleEvent];

  function handleEvents(events: WowEvent[]) {
    events.forEach((event) => {
      if (refTime === -1) {
        refTime = event.time;
      }
      handlers.forEach((handler) => handler(event, state));
    });

    handleStats(state);
  }
  return {
    state,
    handleEvents,
  };
}
