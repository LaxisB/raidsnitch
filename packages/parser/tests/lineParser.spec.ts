import { describe, it } from 'vitest';

import { createParser } from '../src';
import { createLineParser } from '../src/lineParser';
const raw = ``;
const lines = raw.split('\n');

describe('lineParser', () => {
  it('should be defined', ({ expect }) => {
    expect(true).toBe(true);
  });

  it('parses data', async ({ expect }) => {
    const parser = createLineParser(0);
    const res = parser.parseLine('7/1 19:58:19.733  COMBAT_LOG_VERSION,20,ADVANCED_LOG_ENABLED,1,BUILD_VERSION,10.1.0,PROJECT_ID,1');
    expect(res).toMatchSnapshot();
  });
  it('does not throw', async ({ expect }) => {
    try {
      const parser = createLineParser(0);
      lines.forEach((line) => {
        parser.parseLine(line);
      });
    } catch (e) {
      expect.fail(e.message);
    }
  });
});

describe('parser', () => {
  it('parses advanced params', async ({ expect }) => {
    const line =
      '7/16 15:47:57.649  SPELL_DAMAGE_SUPPORT,Player-3686-0A037075,"Montezooma-Antonidas",0x512,0x0,Creature-0-4243-2520-16265-189299-000A33F254,"Decaying Slime",0xa48,0x10,395152,"Ebon Might",0xc,Creature-0-4243-2520-16265-189299-000A33F254,0000000000000000,1058555,1252000,0,0,5043,0,1,0,0,0,-3374.40,4189.57,2096,4.3295,70,4349,2817,-1,1,0,0,0,1,nil,nil,Player-3692-0A0141BE';
    const parser = createParser(0);
    const res = parser.parseLine(line);
    expect(res).toMatchSnapshot();
  });
});
