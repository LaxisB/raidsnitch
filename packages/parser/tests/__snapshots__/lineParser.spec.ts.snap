// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lineParser > parses data 1`] = `
{
  "event": "COMBAT_LOG_VERSION",
  "line": "7/1 19:58:19.733  COMBAT_LOG_VERSION,20,ADVANCED_LOG_ENABLED,1,BUILD_VERSION,10.1.0,PROJECT_ID,1",
  "offset": 0,
  "payload": [
    20,
    "ADVANCED_LOG_ENABLED",
    1,
    "BUILD_VERSION",
    "10.1.0",
    "PROJECT_ID",
    1,
  ],
  "time": 15706699733,
}
`;

exports[`parser > parses advanced params 1`] = `
{
  "advanced": {
    "absorb": 0,
    "armor": 5043,
    "attackPower": 0,
    "currentHp": 1058555,
    "currentPower": 0,
    "facing": 4.3295,
    "infoGuid": "Creature-0-4243-2520-16265-189299-000A33F254",
    "level": 70,
    "maxHp": 1252000,
    "maxPower": 0,
    "ownerGuid": "0000000000000000",
    "positionX": -3374.4,
    "positionY": 4189.57,
    "powerCost": 0,
    "powerType": 1,
    "spellPower": 0,
    "uiMapId": 2096,
  },
  "baseParams": {
    "destFlags": 2632,
    "destGuid": "Creature-0-4243-2520-16265-189299-000A33F254",
    "destName": "Decaying Slime",
    "destRaidFlags": 16,
    "sourceFlags": 1298,
    "sourceGuid": "Player-3686-0A037075",
    "sourceName": "Montezooma-Antonidas",
    "sourceRaidFlags": 0,
  },
  "deltaTime": 0,
  "name": "SPELL_DAMAGE_SUPPORT",
  "prefixes": [
    395152,
    "Ebon Might",
    12,
  ],
  "suffixes": [
    4349,
    2817,
    -1,
    1,
    0,
    0,
    0,
    1,
    null,
    null,
    "Player-3692-0A0141BE",
  ],
  "time": 16987677649,
  "untyped": null,
}
`;
